### Question 5.1: BPE Subword-level Tokenization

Implement subword-level tokenization using a BPE-based tokenizer.
Tokenize the sentence into subword units and display the generated tokens and their corresponding token IDs.

*   **a. Using a pretrained model**
*   **b. Without using a pretrained model**

In [1]:
# Install necessary libraries
!pip install tokenizers transformers -q

In [2]:
# 5.1.a Using Pretrained BPE Model (GPT-2 is BPE-based)
from transformers import GPT2Tokenizer

tokenizer_pretrained = GPT2Tokenizer.from_pretrained('gpt2')

# Read the entire input data file
with open('/content/input_sub_word_data.txt', 'r', encoding='utf-8') as f:
    full_text = f.read().strip()

# Note: the "sequence length > 1024" warning is harmless here, we only tokenize and do not run the model
tokens_pretrained = tokenizer_pretrained.tokenize(full_text)
token_ids_pretrained = tokenizer_pretrained.convert_tokens_to_ids(tokens_pretrained)

print("--- Pretrained GPT-2 BPE Tokenizer (complete input file) ---")
print("Total tokens:", len(tokens_pretrained))
print(f"{'Token':<25} | Token ID")
print("-" * 40)
for t, i in zip(tokens_pretrained, token_ids_pretrained):
    print(f"{t:<25} | {i}")


tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

[transformers] Token indices sequence length is longer than the specified maximum sequence length for this model (1348 > 1024). Running this sequence through the model will result in indexing errors


--- Pretrained GPT-2 BPE Tokenizer (complete input file) ---
Total tokens: 1348
Token                     | Token ID
----------------------------------------
Natural                   | 35364
Ġlanguage                 | 3303
Ġprocessing               | 7587
Ġis                       | 318
Ġa                        | 257
Ġbranch                   | 8478
Ġof                       | 286
Ġartificial               | 11666
Ġintelligence             | 4430
Ġthat                     | 326
Ċ                         | 198
en                        | 268
ables                     | 2977
Ġcomputers                | 9061
Ġto                       | 284
Ġunderstand               | 1833
,                         | 11
Ġinterpret                | 6179
,                         | 11
Ġand                      | 290
Ġgenerate                 | 7716
Ġhuman                    | 1692
Ġlanguage                 | 3303
.                         | 13
Ċ                         | 198
Natural                   | 35

In [3]:
# 5.1.b Train BPE from scratch (Without using a pretrained model)
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import Whitespace

# Initialize a BPE tokenizer
tok_bpe = Tokenizer(BPE(unk_token="[UNK]"))
tok_bpe.pre_tokenizer = Whitespace()

# vocab_size=300: small enough that BPE stops merging before whole words form,
# so reusable subword units (e.g. 'tion', 'ing', 'un') remain visible
trainer = BpeTrainer(special_tokens=["[UNK]", "[PAD]", "[CLS]", "[SEP]", "[MASK]"], vocab_size=300)
tok_bpe.train(["/content/input_sub_word_data.txt"], trainer)
print("Learned vocabulary size:", tok_bpe.get_vocab_size())

# Encode the entire file
encoded = tok_bpe.encode(full_text)

print("\n--- Trained from Scratch BPE Tokenizer (complete input file) ---")
print("Total tokens:", len(encoded.tokens))
print(f"{'Token':<25} | Token ID")
print("-" * 40)
for t, i in zip(encoded.tokens, encoded.ids):
    print(f"{t:<25} | {i}")

# Test on new words suggested in the input data
test_words = ["naturalization", "internationalization", "computationally", "unbelievable",
              "preprocessing", "classification", "communication", "renewable",
              "hydrokinetic", "multilingual", "understanding", "modernization"]
print("\n--- Subword segmentation of test words ---")
for w in test_words:
    e = tok_bpe.encode(w)
    print(f"{w:<22} -> {e.tokens}  IDs: {e.ids}")


Learned vocabulary size: 300

--- Trained from Scratch BPE Tokenizer (complete input file) ---
Total tokens: 2064
Token                     | Token ID
----------------------------------------
N                         | 18
atural                    | 186
language                  | 142
processing                | 192
is                        | 82
a                         | 24
b                         | 25
r                         | 41
an                        | 54
ch                        | 84
of                        | 89
ar                        | 59
t                         | 43
if                        | 114
ic                        | 87
i                         | 32
al                        | 60
in                        | 51
t                         | 43
el                        | 105
l                         | 35
i                         | 32
g                         | 30
ence                      | 178
that                      | 266
en                        

### Question 5.2: SentencePiece Tokenization

Implement subword-level tokenization using a SentencePiece tokenizer with the same input file `/content/input_sub_word_data.txt`.
Tokenize the sentence into subword units and display the generated tokens and their corresponding token IDs.

*   **a. Using a pretrained model**
*   **b. Without using a pretrained model**

In [4]:
# Install sentencepiece library
!pip install sentencepiece -q

In [5]:
# 5.2.a Using Pretrained SentencePiece (ALBERT is SentencePiece-based)
from transformers import AlbertTokenizer

tokenizer_sp_pre = AlbertTokenizer.from_pretrained('albert-base-v2')

with open('/content/input_sub_word_data.txt', 'r', encoding='utf-8') as f:
    full_text = f.read().strip()

# Note: the "sequence length > 512" warning is harmless here, we only tokenize
tokens_sp_pre = tokenizer_sp_pre.tokenize(full_text)
token_ids_sp_pre = tokenizer_sp_pre.convert_tokens_to_ids(tokens_sp_pre)

print("--- Pretrained ALBERT SentencePiece Tokenizer (complete input file) ---")
print("Total tokens:", len(tokens_sp_pre))
print(f"{'Token':<25} | Token ID")
print("-" * 40)
for t, i in zip(tokens_sp_pre, token_ids_sp_pre):
    print(f"{t:<25} | {i}")


tokenizer_config.json:   0%|          | 0.00/25.0 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/760k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.31M [00:00<?, ?B/s]

[transformers] Token indices sequence length is longer than the specified maximum sequence length for this model (1275 > 512). Running this sequence through the model will result in indexing errors


--- Pretrained ALBERT SentencePiece Tokenizer (complete input file) ---
Total tokens: 1275
Token                     | Token ID
----------------------------------------
▁natural                  | 1112
▁language                 | 816
▁processing               | 5511
▁is                       | 25
▁a                        | 21
▁branch                   | 1686
▁of                       | 16
▁artificial               | 6809
▁intelligence             | 2872
▁that                     | 30
▁enables                  | 14645
▁computers                | 7774
▁to                       | 20
▁understand               | 1369
,                         | 15
▁interpret                | 11584
,                         | 15
▁and                      | 17
▁generate                 | 7920
▁human                    | 585
▁language                 | 816
.                         | 9
▁natural                  | 1112
▁language                 | 816
▁processing               | 5511
▁combines                 |

In [6]:
# 5.2.b Train SentencePiece from scratch (Without using a pretrained model)
import sentencepiece as spm

# vocab_size=400: 100 was too small (output was almost character-level);
# 400 lets meaningful subword pieces form while staying within this corpus's limit
spm.SentencePieceTrainer.train(
    input='/content/input_sub_word_data.txt',
    model_prefix='sp_model',
    vocab_size=400,
    model_type='unigram',
    character_coverage=1.0
)

# Load trained model
sp = spm.SentencePieceProcessor(model_file='sp_model.model')

with open('/content/input_sub_word_data.txt', 'r', encoding='utf-8') as f:
    full_text = f.read().strip()

tokens_sp_scratch = sp.encode_as_pieces(full_text)
token_ids_sp_scratch = sp.encode_as_ids(full_text)

print("--- Trained from Scratch SentencePiece (complete input file) ---")
print("Total tokens:", len(tokens_sp_scratch))
print(f"{'Token':<25} | Token ID")
print("-" * 40)
for t, i in zip(tokens_sp_scratch, token_ids_sp_scratch):
    print(f"{t:<25} | {i}")

test_words = ["naturalization", "internationalization", "unbelievable", "hydrokinetic", "multilingual"]
print("\n--- Subword segmentation of test words ---")
for w in test_words:
    print(f"{w:<22} -> {sp.encode_as_pieces(w)}  IDs: {sp.encode_as_ids(w)}")


--- Trained from Scratch SentencePiece (complete input file) ---
Total tokens: 1623
Token                     | Token ID
----------------------------------------
▁Natural                  | 181
▁language                 | 19
▁processing               | 41
▁is                       | 16
▁a                        | 8
▁b                        | 299
ran                       | 381
ch                        | 147
▁of                       | 11
▁arti                     | 287
fic                       | 134
i                         | 38
al                        | 107
▁in                       | 32
te                        | 99
ll                        | 330
i                         | 38
g                         | 398
ence                      | 126
▁that                     | 82
▁en                       | 263
able                      | 63
s                         | 6
▁computer                 | 219
s                         | 6
▁to                       | 14
▁understand            

### Question 5.3: Part-of-Speech (POS) Tagging on Sample Text

Implement Part-of-Speech (POS) tagging on the text: `"The young student is reading an interesting book in the library."` and display the unique token, POS tag, and description.

*   **a. Using SpaCy**
*   **b. Using NLTK**

In [7]:
# Install & download Spacy and NLTK resources
!python -m spacy download en_core_web_sm -q
import nltk
nltk.download('punkt', quiet=True)
nltk.download('punkt_tab', quiet=True)
nltk.download('averaged_perceptron_tagger', quiet=True)
nltk.download('averaged_perceptron_tagger_eng', quiet=True)
nltk.download('tagsets', quiet=True)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.8/12.8 MB 27.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


True

In [8]:
# 5.3.a Using SpaCy
import spacy

nlp = spacy.load("en_core_web_sm")
text = "The young student is reading an interesting book in the library."
doc = nlp(text)

seen_tokens = set()
print(f"{'Token':<15} | {'POS Tag':<8} | {'Description'}")
print("-" * 60)
for token in doc:
    if token.text.lower() not in seen_tokens:
        seen_tokens.add(token.text.lower())
        print(f"{token.text:<15} | {token.pos_:<8} | {spacy.explain(token.pos_)}")

Token           | POS Tag  | Description
------------------------------------------------------------
The             | DET      | determiner
young           | ADJ      | adjective
student         | NOUN     | noun
is              | AUX      | auxiliary
reading         | VERB     | verb
an              | DET      | determiner
interesting     | ADJ      | adjective
book            | NOUN     | noun
in              | ADP      | adposition
library         | NOUN     | noun
.               | PUNCT    | punctuation


In [9]:
# 5.3.b Using NLTK
import nltk
from nltk.tokenize import word_tokenize
from nltk.data import load

# Ensure required NLTK resources are downloaded
nltk.download('punkt_tab', quiet=True)
nltk.download('averaged_perceptron_tagger_eng', quiet=True)
nltk.download('tagsets', quiet=True)
nltk.download('tagsets_json', quiet=True)

# Load official Penn Treebank tag descriptions (covers every tag, no hardcoding)
try:
    tagdict = load('help/tagsets_json/PY3_json/upenn_tagset.json')
except Exception:
    tagdict = load('help/tagsets/upenn_tagset.pickle')

text = "The young student is reading an interesting book in the library."
tokens = word_tokenize(text)
pos_tags = nltk.pos_tag(tokens)

seen_tokens = set()
print(f"{'Token':<15} | {'POS Tag':<8} | {'Description'}")
print("-" * 60)
for token, tag in pos_tags:
    if token.lower() not in seen_tokens:
        seen_tokens.add(token.lower())
        desc = tagdict.get(tag, ('unknown',))[0]
        print(f"{token:<15} | {tag:<8} | {desc}")


Token           | POS Tag  | Description
------------------------------------------------------------
The             | DT       | determiner
young           | JJ       | adjective or numeral, ordinal
student         | NN       | noun, common, singular or mass
is              | VBZ      | verb, present tense, 3rd person singular
reading         | VBG      | verb, present participle or gerund
an              | DT       | determiner
interesting     | JJ       | adjective or numeral, ordinal
book            | NN       | noun, common, singular or mass
in              | IN       | preposition or conjunction, subordinating
library         | NN       | noun, common, singular or mass
.               | .        | sentence terminator


### Question 5.4: POS Tagging & Frequency using SpaCy

Implement Part-of-Speech (POS) tagging on the given file `/content/input_sub_word_data.txt` and display the unique token, POS tag, description, and frequency using SpaCy.

In [10]:
from collections import Counter
import spacy

nlp = spacy.load("en_core_web_sm")

# Read all text from file
with open('/content/input_sub_word_data.txt', 'r', encoding='utf-8') as f:
    large_text = f.read()

doc = nlp(large_text)

# Frequency of each (token, POS tag) pair, so a word used with two different
# POS tags gets a separate, correct count for each
pair_freq = Counter((token.text.lower(), token.pos_) for token in doc if not token.is_space)

print(f"Total unique (token, POS) entries: {len(pair_freq)}\n")
print(f"{'Token':<22} | {'POS Tag':<8} | {'Frequency':<10} | {'Description'}")
print("-" * 75)

for (tok, pos), freq in pair_freq.most_common():
    print(f"{tok:<22} | {pos:<8} | {freq:<10} | {spacy.explain(pos)}")


Total unique (token, POS) entries: 395

Token                  | POS Tag  | Frequency  | Description
---------------------------------------------------------------------------
,                      | PUNCT    | 76         | punctuation
.                      | PUNCT    | 65         | punctuation
the                    | DET      | 47         | determiner
a                      | DET      | 28         | determiner
and                    | CCONJ    | 27         | coordinating conjunction
can                    | AUX      | 23         | auxiliary
of                     | ADP      | 22         | adposition
be                     | AUX      | 15         | auxiliary
is                     | AUX      | 14         | auxiliary
as                     | ADP      | 13         | adposition
bpe                    | PROPN    | 13         | proper noun
language               | NOUN     | 12         | noun
word                   | NOUN     | 12         | noun
tokenizer              | NOUN     | 11   

### Question 5.5: Train Word2Vec CBOW Model

Create a file named `corpus.txt` with the given input corpus, train a Word2Vec CBOW model, and find the three most similar words to a selected word.

In [11]:
# Create corpus.txt
corpus_content = """Natural language processing is interesting
Natural language processing is useful
I love machine learning
Machine learning is useful
Deep learning is a part of machine learning
Artificial intelligence is changing the world
Natural language processing uses machine learning
Deep learning and artificial intelligence are related
I study natural language processing
Machine learning helps artificial intelligence"""

with open('corpus.txt', 'w', encoding='utf-8') as f:
    f.write(corpus_content)

In [12]:
# Install gensim library for Word2Vec
!pip install gensim -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.8/27.8 MB 18.5 MB/s eta 0:00:00


In [13]:
from gensim.models import Word2Vec

# Load and preprocess corpus
sentences = []
with open('corpus.txt', 'r', encoding='utf-8') as f:
    for line in f:
        if line.strip():
            sentences.append(line.strip().lower().split())

# Hyperparameters tuned for this very small (10-sentence) corpus:
#   vector_size=20 : few words, so small vectors avoid too many free parameters
#   window=2       : sentences are short (4-8 words), a wide window mixes unrelated words
#   epochs=300     : few examples, so more passes are needed to learn anything
#                    (500+ over-trains: every word becomes ~0.98 similar to every other)

# Train CBOW model (sg=0 represents CBOW); seed + workers=1 make results reproducible
cbow_model = Word2Vec(sentences, vector_size=20, window=2, min_count=1, sg=0,
                      epochs=300, seed=42, workers=1)

# Selected word representation
selected_word = "learning"
print(f"Vector Representation of '{selected_word}':\n", cbow_model.wv[selected_word])

# Find three most similar words
similar_words = cbow_model.wv.most_similar(selected_word, topn=3)
print(f"\nThree most similar words to '{selected_word}':")
for w, s in similar_words:
    print(f"  {w:<15} {s:.4f}")


Vector Representation of 'learning':
 [ 0.03931578  0.06480844 -0.00854276  0.05461651  0.02679948  0.09169175
 -0.12730786  0.06043514 -0.05603912 -0.10599907  0.02184465  0.02356905
  0.00753    -0.00542104  0.11360407 -0.00616522 -0.00156152 -0.11588403
  0.01544915 -0.01613381]

Three most similar words to 'learning':
  intelligence    0.8711
  deep            0.8447
  are             0.7988


### Question 5.6: Train Word2Vec Skip-Gram Model

Train a Skip-Gram Word2Vec model using the same `corpus.txt`. Retrieve the vector for `"language"` and display the five most similar words.

In [14]:
# Train Skip-Gram model (sg=1 represents Skip-Gram)
sg_model = Word2Vec(sentences, vector_size=20, window=2, min_count=1, sg=1,
                    epochs=300, seed=42, workers=1)

# Selected word representation
selected_word_sg = "language"
print(f"Vector Representation of '{selected_word_sg}':\n", sg_model.wv[selected_word_sg])

# Find five most similar words
similar_words_sg = sg_model.wv.most_similar(selected_word_sg, topn=5)
print(f"\nFive most similar words to '{selected_word_sg}':")
for w, s in similar_words_sg:
    print(f"  {w:<15} {s:.4f}")


Vector Representation of 'language':
 [ 0.08330168  0.0149668  -0.04751608  0.06763475  0.03221443  0.04878259
 -0.08518738  0.00452707  0.01282248 -0.00281574  0.00538134  0.01520111
  0.02419697 -0.02805933  0.09098046 -0.03763697 -0.02825882 -0.02068557
  0.0018713  -0.04132954]

Five most similar words to 'language':
  a               0.8083
  artificial      0.7927
  the             0.7821
  intelligence    0.7415
  deep            0.7304


### Question 5.7: Word2Vec Cosine Similarity

Train a Word2Vec model using the provided corpus and calculate the cosine similarity between selected word pairs:

*   machine and learning
*   natural and language
*   deep and learning

In [15]:
# Train Word2Vec model
w2v_model = Word2Vec(sentences, vector_size=20, window=2, min_count=1, sg=1,
                     epochs=300, seed=42, workers=1)

pairs = [
    ('machine', 'learning'),
    ('natural', 'language'),
    ('deep', 'learning')
]

for w1, w2 in pairs:
    similarity = w2v_model.wv.similarity(w1, w2)
    print(f"Cosine similarity between '{w1}' and '{w2}': {similarity:.4f}")

# Baseline: an unrelated pair, to show the model separates related from unrelated words
print(f"\n(Baseline) 'language' vs 'world': {w2v_model.wv.similarity('language', 'world'):.4f}")


Cosine similarity between 'machine' and 'learning': 0.7860
Cosine similarity between 'natural' and 'language': 0.6920
Cosine similarity between 'deep' and 'learning': 0.8942

(Baseline) 'language' vs 'world': 0.4455


**Note on Word2Vec results (5.5–5.7):** The experiment specifies this 10-sentence `corpus.txt`, so it is used as given. Because the corpus is tiny, the hyperparameters were tuned for it (vector_size=20, window=2, epochs=300). The related pairs score about 0.7–0.9, well above the unrelated baseline (language–world, about 0.45). Common words such as 'a', 'is' and 'the' may still appear among the nearest neighbours, because in such a small corpus they share contexts with almost every word.